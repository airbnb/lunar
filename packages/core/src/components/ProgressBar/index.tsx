import React from 'react';
import { between } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Disable leading rounded corners. */
  leading?: boolean;
  /** Percent in which the progess is complete, ranging from 0 to 100. */
  percent: number;
  /** Disable trailing rounded corners. */
  trailing?: boolean;
};

/** A bar to represent the progress to completion. */
export class ProgressBar extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    percent: between({ gte: 0, lte: 100 }).isRequired,
  };

  static defaultProps = {
    leading: false,
    trailing: false,
  };

  render() {
    const { cx, percent, leading, trailing, styles } = this.props;

    return (
      <div className={cx(styles.wrapper)}>
        <div
          className={cx(
            styles.bar,
            !leading && styles.bar_leading,
            !trailing && styles.bar_trailing,
          )}
        >
          <div
            className={cx(
              styles.bar,
              !leading && styles.bar_leading,
              !trailing && styles.bar_trailing,
              styles.progress,
            )}
            style={{ width: `${percent}%` }}
          />
        </div>
      </div>
    );
  }
}

export default withStyles(({ color, unit, ui }) => ({
  wrapper: {
    paddingTop: unit / 4,
    paddingBottom: unit / 4,
  },

  bar: {
    height: unit / 2,
    background: color.core.primary[1],
  },

  bar_leading: {
    borderTopRightRadius: ui.borderRadius,
    borderBottomRightRadius: ui.borderRadius,
  },

  bar_trailing: {
    borderTopLeftRadius: ui.borderRadius,
    borderBottomLeftRadius: ui.borderRadius,
  },

  progress: {
    background: color.core.primary[6],
  },
}))(ProgressBar);
