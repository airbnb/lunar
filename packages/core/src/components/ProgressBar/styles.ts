import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, ui, unit }) => ({
  wrapper: {
    paddingTop: unit / 4,
    paddingBottom: unit / 4,
  },

  bar: {
    height: unit / 2,
    background: color.core.primary[1],
  },

  bar_leading: {
    borderTopRightRadius: ui.borderRadius,
    borderBottomRightRadius: ui.borderRadius,
  },

  bar_trailing: {
    borderTopLeftRadius: ui.borderRadius,
    borderBottomLeftRadius: ui.borderRadius,
  },

  progress: {
    background: color.core.primary[6],
  },
});

export default styleSheet;

export { styleSheet };
