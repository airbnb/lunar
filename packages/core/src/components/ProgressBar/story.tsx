import React from 'react';
import ProgressBar from '.';

export default {
  title: 'Core/ProgressBar',
  parameters: {
    inspectComponents: [ProgressBar],
  },
};

export function aProgressBarWithDifferentCompletionPercentagesAndWidths() {
  return (
    <>
      <ProgressBar percent={1} />
      <br />
      <ProgressBar percent={29} />
      <br />
      <ProgressBar percent={100} />
      <br />
      <ProgressBar percent={63} />
    </>
  );
}

aProgressBarWithDifferentCompletionPercentagesAndWidths.story = {
  name: 'A progress bar with different completion percentages and widths.',
};

export function canDisableLeadingAndTrailingEdgesNoRoundedCorners() {
  return (
    <>
      <ProgressBar leading percent={50} />
      <br />
      <ProgressBar trailing percent={50} />
    </>
  );
}

canDisableLeadingAndTrailingEdgesNoRoundedCorners.story = {
  name: 'Can disable leading and trailing edges (no rounded corners).',
};
