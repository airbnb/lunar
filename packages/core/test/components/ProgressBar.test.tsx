import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import ProgressBar from '../../src/components/ProgressBar';

describe('<ProgressBar />', () => {
  it('renders a progress bar', () => {
    const wrapper = shallowWithStyles(<ProgressBar percent={10}>Button</ProgressBar>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders trailing', () => {
    const wrapper = shallowWithStyles(
      <ProgressBar trailing percent={50}>
        Button
      </ProgressBar>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders leading', () => {
    const wrapper = shallowWithStyles(
      <ProgressBar leading percent={50}>
        Button
      </ProgressBar>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can change width with percent prop', () => {
    const wrapper = shallowWithStyles(<ProgressBar percent={15}>Button</ProgressBar>);

    expect(
      wrapper
        .find('div')
        .at(2)
        .prop('style'),
    ).toEqual({ width: '15%' });

    wrapper.setProps({
      percent: 85,
    });

    expect(
      wrapper
        .find('div')
        .at(2)
        .prop('style'),
    ).toEqual({ width: '85%' });
  });
});
