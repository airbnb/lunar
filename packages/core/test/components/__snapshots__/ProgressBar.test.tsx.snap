// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> renders a progress bar 1`] = `
<div
  className="wrapper_fyg7bf"
>
  <div
    className="bar_xmolfv-o_O-bar_leading_satysr-o_O-bar_trailing_j3624r"
  >
    <div
      className="bar_xmolfv-o_O-bar_leading_satysr-o_O-bar_trailing_j3624r-o_O-progress_hk8cyc"
      style={
        Object {
          "width": "10%",
        }
      }
    />
  </div>
</div>
`;

exports[`<ProgressBar /> renders leading 1`] = `
<div
  className="wrapper_fyg7bf"
>
  <div
    className="bar_xmolfv-o_O-bar_trailing_j3624r"
  >
    <div
      className="bar_xmolfv-o_O-bar_trailing_j3624r-o_O-progress_hk8cyc"
      style={
        Object {
          "width": "50%",
        }
      }
    />
  </div>
</div>
`;

exports[`<ProgressBar /> renders trailing 1`] = `
<div
  className="wrapper_fyg7bf"
>
  <div
    className="bar_xmolfv-o_O-bar_leading_satysr"
  >
    <div
      className="bar_xmolfv-o_O-bar_leading_satysr-o_O-progress_hk8cyc"
      style={
        Object {
          "width": "50%",
        }
      }
    />
  </div>
</div>
`;
